chore(docs): add content sync to source plugin author guide (#34250)
gatsbyjs/gatsby: docs/docs/how-to/plugins-and-themes/creating-a-source-plugin.md

An easy way to keep track of your manifest logic is to parse it out into a different util function. Either inside the `createNodeManifest` util or before you call it you'll need to vet which nodes you'll want to create manifests for.

```javascript:title=source-plugin/gatsby-node.js
import { createNodeManifest } from "./utils.js"
exports.sourceNodes = async (
  { actions }
) => {
  // sourcing data...
  const { unstable_createNodeManifest, createNode } = actions

  nodes.forEach(node => {
    // highlight-start
    const gatsbyNode = createNode(node)

    const nodeIsEntryNode = `some condition`
    if (nodeIsEntryNode) {
      createNodeManifest({
        entryItem: node,
        entryNode: gatsbyNode,
        project,
        unstable_createNodeManifest,
      })
    }
    // highlight-end
  })
}
```

> There are a couple more checks we need to do inside the util. First, at the moment you'll only want to create node manifests for preview content, next we need to check if the Gatsby version supports `unstable_createNodeManifest`. Finally we'll want to parse out the time the content was updated and create a `manifestId` with all that information. The manifest id is the link between the data source and the Gatsby site. This manifest id will need to be created identically in the data source and here in your source plugin. Because of this we will use information given from the data source to build up the manifest id.

Reviewer: I think this could be easier to parse out with subheadings for each thing needed to implement it.

Reviewer 2: I like the idea of numbered steps to make it easier to follow, but the content is great 👍

Some data sources keep event logs and are able to return a list of objects modified since a given time. If you're building a source plugin, you can store the last time you fetched data using the [cache](/docs/creating-a-source-plugin/#caching-data-between-runs) and then only sync down nodes that have been modified since that time. [`gatsby-source-contentful`](https://github.com/gatsbyjs/gatsby/tree/master/packages/gatsby-source-contentful) is an example of a source plugin that does this.

### Enabling Content Sync

> If you would like to add Content Sync to your source plugin here but aren't sure what it is [learn more about Content Sync here](https://www.gatsbyjs.com/docs/conceptual/content-sync/). To enable this feature in your source plugin you will need to make sure that your data source (or CMS) also works with Content Sync. The first thing you'll want to do is identify which nodes you'll want to create a node manifest for. These will typically be nodes that you can preview, entry nodes, top level nodes, etc. An example of this could be a blog post or an article, any node that can be the "owner" of a page. Once you've identified that you'll want to use the [unstable_createNodeManifest action](https://www.gatsbyjs.com/docs/reference/config-files/actions/#unstable_createNodeManifest) to create these manifests. A good place to call this action is before you call `createNode`.

Author: We'll probably also want a guide on how to implement this on the CMS side. I don't believe we have this yet

Reviewer 3: Doc links need to be relative

Reviewer: Yeah, if we're mentioning how to implement it in source plugins we need the CMS side as well for it to make sense. This is awesome though :)

Some data sources keep event logs and are able to return a list of objects modified since a given time. If you're building a source plugin, you can store the last time you fetched data using the [cache](/docs/creating-a-source-plugin/#caching-data-between-runs) and then only sync down nodes that have been modified since that time. [`gatsby-source-contentful`](https://github.com/gatsbyjs/gatsby/tree/master/packages/gatsby-source-contentful) is an example of a source plugin that does this.

### Enabling Content Sync

If you would like to add Content Sync to your source plugin here but aren't sure what it is [learn more about Content Sync here](/docs/conceptual/content-sync/). To enable this feature in your source plugin you will need to make sure that your data source (or CMS) also works with Content Sync.

#### Content Sync Source Plugin changes

The source plugin needs to create node manifests using the [`unstable_createNodeManifest` action](/docs/reference/config-files/actions/#unstable_createNodeManifest).

##### Identifying which nodes need manifests

> 

Reviewer 4: We could also just link to the official docs for the api :point_up:

> 

Author: Yeah it's already linking to https://www.gatsbyjs.com/docs/reference/config-files/actions/#unstable_createNodeManifest that seems like enough explanation?

> 

Reviewer 4: Haha, I totally glossed over that link :sweat_smile: Looks good to me :heavy_check_mark:

export function createNodeManifest({
  entryItem, // the raw data source/cms content data
  project,   // the cms project data
  entryNode, // the Gatsby node
  unstable_createNodeManifest,
}) {
  // highlight-start
  // This env variable is provided automatically on Gatsby Cloud hosting
  const isPreview = process.env.GATSBY_IS_PREVIEW === `true`

  const createNodeManifestIsSupported =
    typeof unstable_createNodeManifest === `function`

  const shouldCreateNodeManifest = isPreview && createNodeManifestIsSupported
  // highlight-end

  if (shouldCreateNodeManifest) {
    // create manifest...
  }
}
```

##### Call `unstable_createNodeManifest`

Next we will build up the `manifestId` and call `unstable_createNodeManifest`. The `manifestId` needs to be created with information that comes from the data source NOT Gatsby, which is why we use the `entryItem` id as opposed to the `entryNode` id. This `manifestId` must be uniquely tied to a specific revision of specific content. We use the CMS project space (you may not need this), the id of the content, and finally the timestamp that it was updated at.

```javascript:title=source-plugin/utils.js

> // highlight-start

Author: This is to highlight code to direct attention to specific spots in the code block

> // highlight-start

Reviewer 4: Totally! But this variable seems unrelated to the text above and not necessarily something we want to highlight in this exact section

> // highlight-start

Author: oh whoops! This was leftover from copy and pasting, 881-883 shoud be deleted

let warnOnceForNoSupport = false
// highlight-end

export function createNodeManifest({
  // ...
}) {
  // ...

  if (shouldCreateNodeManifest) {

>     // highlight-start

Reviewer 4: Oops, I meant to add a suggestion to add a `highlight-end` comment :face_exhaling:

>     // highlight-start

Author: Aha yes!

    unstable_createNodeManifest({
      manifestId,
      node: entryNode,
      updatedAtUTC: updatedAt,
    })
  }
}
```

##### Warn if no support

Lastly we'll want to give our users a good experience and give a warning if they're using a version of Gatsby that does not support Content Sync

```javascript:title=source-plugin/utils.js
// highlight-start
let warnOnceForNoSupport = false
// highlight-end

export function createNodeManifest({
  // ...
}) {
  // ...

  if (shouldCreateNodeManifest) {
    // ...
  } else if (
    // highlight-start
    // it's helpful to let users know if they're using an outdated Gatsby version so they'll upgrade for the best experience
    isPreview && !createNodeManifestIsSupported && !warnOnceForNoSupport
  ) {
    console.warn(
      `${sourcePluginName}: Your version of Gatsby core doesn't support Content Sync (via the unstable_createNodeManifest action). Please upgrade to the latest version to use Content Sync in your site.`
    )
    // This is getting called for every entry node so we don't want the console logs to get cluttered
    warnOnceForNoSupport = true
    // highlight-end
  }
}
```

#### Content Sync data source (or CMS) changes

> The CMS will need to send a webhook when content is changed

Reviewer 4: ```suggestion
The CMS will need to send a preview webhook to Gatsby Cloud when content is changed. Find out how to get that webhook [here](https://support.gatsbyjs.com/hc/en-us/articles/360052324394-Build-and-Preview-Webhooks)
```

Author: oops I forgot to finish this sentence 😆

##### Using the gatsby preview extension

> This package is currently in the process of being open sourced and should be available shortly. For context, here is what the package provides, a button to handle opening the Content Sync waiting room. You will need to create a button similar to this for the time being.

Reviewer 4: ```suggestion
We are currently in process of creating an open source package to handle most of this functionality for you. In the meantime, you will need to create a button in your CMS that does the following...
1.  `POST` to the preview webhook url in Gatsby cloud
2. Open the Content Sync waiting room

The button might look something like this:
```

![Open Preview Button Screenshot](https://user-images.githubusercontent.com/18426780/61498855-37e35a00-a982-11e9-8201-ab96be74f1f1.png)

##### Configuration

> You will need to store the Content Sync URL from a given Gatsby Cloud site. This will look something like `https://gatsbyjs.com/content-sync/<siteId>`. This is often done in the plugin extension or app extension configuration, this will differ from CMS to CMS depending on how they handle their plugin ecosystem.

Reviewer 4: Do we have a doc explaining where to grab a site's content sync url we can link to here?

Author: Hmm, i don't think so, but i'll just mention that it's the exact same place you find your webhook urls

##### Building up the manifest id

Recall that we need to create a matching manifest id in the CMS AND the Gatsby plugin. Whenever content is saved we can build up a new manifest id that will look the same as the manifest id we created in the source plugin

```javascript

> const manifestId = `${project.id}-${entryItem.id}-${updatedAt}`

Reviewer 4: A small suggestion to make this a little more generic. `entryItem` is Contentful specific and also might imply some sort of top level node in a tree which might be a tad confusing if you aren't thinking about Contentful already :slightly_smiling_face: 

```suggestion
const manifestId = `${project.id}-${blogPost.id}-${updatedAt}`
```

Author: I already explain the context of an entryItem earlier when I talk through source plugins and the idea of an entry node is used repeatedly throughout, this is code copied directly from there. This was to demonstrate that it will be the exact same code in both source plugin and CMS to build up the manifestId

> The first thing you'll want to do is identify which nodes you'll want to create a node manifest for. These will typically be nodes that you can preview, entry nodes, top level nodes, etc. An example of this could be a blog post or an article, any node that can be the "owner" of a page. A good place to call this action is whenever you call `createNode`.

Reviewer 4: Aha, got it. It makes sense to leave as is :white_check_mark:

```

> In the CMS we should have access to the project id (if the CMS uses one) and the content id, the only thing we need to figure out how to grab is the timestamp that the content was updated at.

Reviewer 4: ```suggestion
In the CMS plugin, we should have access to
- the project id (if the CMS uses one)
- the content id
- the timestamp that the content was updated at (or some other piece of data that is tied to a very specific state of saved content)
```

> 

Reviewer 4: ```suggestion
##### Starting the preview build

Next, we need to tell Gatsby cloud to build a preview by `POST`ing to the Gatsby cloud preview build webhook. 
```

##### Opening the Content Sync waiting room

> When we've built up the manifest id, when the "Open Preview" button is clicked, we need to open a new window to the Content Sync waiting room as follows `https://gatsbyjs.com/content-sync/<siteId>/<sourcePluginName>/<manifestId>`. This will ensure that this piece of content is linked to the corresponding content that was created in their Gatsby site.

Reviewer 4: ```suggestion
Once we've built a `manifestId` and `POST`ed to the preview build webhook, we need to open a new tab/window with a modified version of the Content Sync URL. You get that by grabbing the Content Sync URL you stored in the CMS plugin earlier and appending the Gatsby source plugin name and content's `manifestId` that you just created. Like so...
`https://gatsbyjs.com/content-sync/<siteId>/<sourcePluginName>/<manifestId>`
```

I'm mainly try to call out the connection between this url and the stored content sync url from an earlier step with this suggestion

#### Useful Content Sync development tips

> There are a couple gotchas depending on how the CMS acts. Sometimes you will need to wait to make sure you have the correct `updatedAt` timestamp as some CMS may take a second to update their backend and then wait for the change to propagate to the frontend. While others will immediately update the frontend and then propagate that to the backend.

Reviewer 4: ```suggestion
There are a couple "gotchas" depending on how the CMS acts.

- Sometimes you will need to wait to make sure you have the correct `updatedAt` timestamp as some CMS may take a second to update their backend and then wait for the change to propagate to the frontend. While others will immediately update the frontend and then propagate that to the backend.
```

> Make sure that a preview webhook is being sent to Gatsby Cloud after the content is edited, whether it's before you press the "Open Preview" button or the "Open Preview" is the trigger that sends the webhook.

Reviewer 4: ```suggestion
- Make sure that a preview webhook is being sent to Gatsby Cloud after the content is edited, whether it's before you press the "Open Preview" button or the "Open Preview" is the trigger that sends the webhook.
```

> While developing, you can set the Gatsby `VERBOSE` env variable to `"true"` to see additional logs that will help you debug what's happening in the source plugin. Ideally when you click the `Open Preview` button in the CMS the `manifestId` in the URL should match the `manifestId` that the source plugin creates from that revision.

Reviewer 4: Breaking up all of the unique items into separate bullets for easier scanning and more clear organization...

```suggestion
- While developing, you can set the Gatsby `VERBOSE` env variable to `"true"` to see additional logs that will help you debug what's happening in the source plugin.
- Ideally when you click the `Open Preview` button in the CMS the `manifestId` in the URL should match the `manifestId` that the source plugin creates from that revision.
```

Reviewer 4: I'm a little confused what you mean by this? Doesn't this 100% need to be the case in order for Content Sync to work?

> Ideally when you click the `Open Preview` button in the CMS the `manifestId` in the URL should match the `manifestId` that the source plugin creates from that revision.

Author: It's the happy path 😆 , yeah I'll change that to make it sound more like that's the only way it works

Reviewer 4: Okay sweet, thanks for clarifying! (and making the tweak) :smile:

  ) {
    console.warn(
      `${sourcePluginName}: Your version of Gatsby core doesn't support Content Sync (via the unstable_createNodeManifest action). Please upgrade to the latest version to use Content Sync in your site.`
    )
    // This is getting called for every entry node so we don't want the console logs to get cluttered
    warnOnceForNoSupport = true
    // highlight-end
  }
}
```

#### Content Sync data source (or CMS) changes

The CMS will need to send a preview webhook to Gatsby Cloud when content is changed and open the Content Sync waiting room. Follow along to learn how to implement it on the CMS side.

##### Using the Gatsby preview extension

We are currently in process of creating an open source package to handle most of this functionality for you. In the meantime, you will need to create a button in your CMS that does the following...

1. `POST` to the preview webhook url in Gatsby Cloud
2. Open the Content Sync waiting room

The button might look something like this:
![Open Preview Button Screenshot](https://user-images.githubusercontent.com/18426780/61498855-37e35a00-a982-11e9-8201-ab96be74f1f1.png)

##### Configuration

> You will need to store the Content Sync URL from a given Gatsby Cloud site. This will look something like `https://gatsbyjs.com/content-sync/<siteId>`. This is often done in the CMS plugin extension configuration, this will differ from CMS to CMS depending on how they handle their plugin ecosystem.

Reviewer: We should mention that this needs to be user-configurable. So stored in a CMS options page or something like that.

```javascript
const manifestId = `${project.id}-${entryItem.id}-${updatedAt}`
```

In the CMS extension, we should have access to

- the project id (if the CMS uses one)
- the content id
- the timestamp that the content was updated at (or some other piece of data that is tied to a very specific state of saved content)

##### Starting the preview build (optional)

If the CMS does not handle this part automatically we will need to tell Gatsby cloud to build a preview by `POST`ing to the Gatsby Cloud preview build webhook url.

##### Opening the Content Sync waiting room

Once we've built a `manifestId` and `POST`ed to the preview build webhook url, we need to open a new tab/window with a modified version of the Content Sync URL. You get that by grabbing the Content Sync URL you stored in the CMS extension earlier and appending the Gatsby source plugin name and the content's `manifestId` that you just created, `https://gatsbyjs.com/content-sync/<siteId>/<sourcePluginName>/<manifestId>`.

#### Useful Content Sync development tips

Here are some things to keep in mind and some "gotchas" depending on how the CMS acts.

> - Sometimes you will need to wait to make sure you have the correct `updatedAt` timestamp as some CMS may take a second to update their backend and then wait for the change to propagate to the frontend. While others will immediately update the frontend and then propagate that to the backend.

Reviewer: It'd be good to mention this is within the CMS, it's not 100% clear where we're waiting

- Make sure that a preview webhook is being sent to Gatsby Cloud after the content is edited, whether it's before you press the "Open Preview" button or the "Open Preview" is the trigger that sends the webhook.

> - While developing, you can set the Gatsby `VERBOSE` env variable to `"true"` to see additional logs that will help you debug what's happening in the source plugin.

Reviewer: Is this different from `gatsby develop --verbose` or the same thing?

Author: it's the same thing, gatsby checks for either and then enables verbose logging

Reviewer: Cool, good to know 👍 that's what I thought but I wasn't 100% sure

Reviewer: Thanks!

- When you click the "Open Preview" button in the CMS the `manifestId` in the URL should match the `manifestId` that the source plugin creates from that revision.

> - The node manifests get written out in the `public` dir of your gatsby site, so you can check to manifests on your local disk `/public/<sourcePluginName>/<manifestId>` or you can navigate directly to that piece of content `https://<your-domain>/<sourcePluginName>/<manifestId>`

Reviewer: It should be `/public/__node-manifests/<sourcePluginName>/<manifestId>.json` and `https://<your-domain>/__node-manifests/<sourcePluginName>/<manifestId>`
